dht.go

```diff
@@ -507,3 +578,46 @@ func (dht *IpfsDHT) newContextWithLocalTags(ctx context.Context, extraTags ...ta
 	) // ignoring error as it is unrelated to the actual function of this code.
 	return ctx
 }
+
+func (dht *IpfsDHT) handleProtocolChanges(ctx context.Context) {
+	// register for event bus protocol ID changes
+	sub, err := dht.host.EventBus().Subscribe(new(event.EvtPeerProtocolsUpdated))
+	if err != nil {
+		panic(err)
+	}
+	defer sub.Close()
+
+	pmap := make(map[protocol.ID]bool)
+	for _, p := range dht.protocols {
+		pmap[p] = true
+	}
+
+	for {
+		select {
+		case ie, ok := <-sub.Out():
+			e, ok := ie.(event.EvtPeerProtocolsUpdated)
+			if !ok {
+				logger.Errorf("got wrong type from subscription: %T", ie)
+				return
+			}
+
+			protos, err := dht.peerstore.SupportsProtocols(e.Peer, dht.protocolStrs()...)
+			if err != nil {
+				continue
+			}
+
+			if len(protos) > 0 {
+				for _, p := range e.Added {
+					if pmap[p] {
+						dht.routingTable.Update(e.Peer)
+						break
+					}
+				}
```

Comment:
@Stebalien @aarshkshah1992 I added in the check for `e.Added` because `dht.routingTable.Update()` is stateful and pushes the peer to the front of the bucket. AFAIK we either are or will be changing this interface. Will the replacement call here be idempotent, if so I can remove the `e.Added` check either now or after the routing table changes land.

Reply:
The new method should be idempotent. We basically just say "this peer exists" and the routing table reacts.

Reply:
@aschmahmann I don't understand why you want to remove this call. If we connect to a new peer that's not listening to the DHT protocol, we wont have it in the RT.

However, if it later starts listening to the DHT protocol, shouldn't we add it to the RT here ?

Reply:
@aschmahmann You mean you will add the peer to the RT if the peerstore says it supports the protocol without checking if the peer has newly added support for it, right ?

That should work fine.